File: headless/src/server/teable.ts
import { getSettings, saveSetting } from "./database";
import type { Lap, LapSession, TeableUser, TeableSpace, TeableBase, TeableTable, TeableSchemaCheck } from "../shared/types";

// ─── Required fields for the Laps table ──────────────────────────────────────

interface RequiredField {
  name: string;
  type: string;
}

const REQUIRED_FIELDS: RequiredField[] = [
  { name: "Session Started At", type: "singleLineText" },
  { name: "Lap Mode",           type: "singleLineText" },
  { name: "Lap Number",         type: "number" },
  { name: "Start Time",         type: "singleLineText" },
  { name: "End Time",           type: "singleLineText" },
  { name: "Duration",           type: "singleLineText" },
  { name: "Speed At Start",     type: "number" },
  { name: "Speed At End",       type: "number" },
];

// ─── Helpers ──────────────────────────────────────────────────────────────────

const FETCH_OPTS = { tls: { rejectUnauthorized: false } } as RequestInit;

function buildHeaders(token: string): Record<string, string> {
  return {
    "Authorization": `Bearer ${token}`,
    "Content-Type": "application/json",
  };
}

function apiFetch(url: string, init?: RequestInit): Promise<Response> {
  return fetch(url, { ...FETCH_OPTS, ...init, headers: { ...init?.headers } });
}

function formatDuration(ms: number): string {
  const total = Math.round(ms);
  const totalSeconds = Math.floor(total / 1000);
  const minutes = Math.floor(totalSeconds / 60);
  const seconds = totalSeconds % 60;
  const millis = total % 1000;
  return `${String(minutes).padStart(2, "0")}:${String(seconds).padStart(2, "0")}.${String(millis).padStart(3, "0")}`;
}

function getConfig(): { url: string; token: string } {
  const settings = getSettings();
  const url = settings.teableUrl.replace(/\/+$/, "");
  const token = settings.teableToken;
  if (!url) throw new Error("Teable not configured");
  if (!token) throw new Error("Teable token not found");
  return { url, token };
}

// ─── Connection test ──────────────────────────────────────────────────────────

export async function testTeableConnection(url: string, token: string): Promise<TeableUser> {
  const base = url.replace(/\/+$/, "");

  let spacesResp: Response;
  try {
    spacesResp = await apiFetch(`${base}/api/space`, { headers: buildHeaders(token) });
  } catch (e) {
    throw new Error(`Could not connect to ${base} — check the URL and your network connection. Details: ${String(e)}`);
  }

  if (spacesResp.status === 401) {
    throw new Error("Invalid or expired access token. Please check your Personal Access Token.");
  }
  if (!spacesResp.ok) {
    const body = await spacesResp.text().catch(() => "");
    throw new Error(`Teable API returned ${spacesResp.status} — ${body}`);
  }

  let user: TeableUser = { id: "", name: "Teable User", email: "", avatar: null };
  try {
    const userResp = await apiFetch(`${base}/api/auth/user/me`, { headers: buildHeaders(token) });
    if (userResp.ok) {
      const data = await userResp.json() as { id?: string; name?: string; email?: string; avatar?: string };
      user = {
        id: data.id ?? "",
        name: data.name ?? "Teable User",
        email: data.email ?? "",
        avatar: data.avatar ?? null,
      };
    }
  } catch {
    // Fallback
  }

  return user;
}

// ─── Config management ────────────────────────────────────────────────────────

export function saveTeableConfig(
  url: string,
  token: string,
  userName: string,
  userEmail: string,
  userAvatar: string | null,
): void {
  saveSetting("teableUrl", url);
  saveSetting("teableToken", token);
  saveSetting("teableUserName", userName);
  saveSetting("teableUserEmail", userEmail);
  saveSetting("teableUserAvatar", userAvatar ?? "");
}

export function removeTeableConfig(): void {
  saveSetting("teableUrl", "");
  saveSetting("teableToken", "");
  saveSetting("teableUserName", "");
  saveSetting("teableUserEmail", "");
  saveSetting("teableUserAvatar", "");
  saveSetting("teableSpaceId", "");
  saveSetting("teableBaseId", "");
  saveSetting("teableTableId", "");
  saveSetting("teableSyncEnabled", "false");
}

export function saveTeableTarget(spaceId: string, baseId: string, tableId: string): void {
  saveSetting("teableSpaceId", spaceId);
  saveSetting("teableBaseId", baseId);
  saveSetting("teableTableId", tableId);
}

// ─── Spaces / Bases / Tables ──────────────────────────────────────────────────

export async function listTeableSpaces(): Promise<TeableSpace[]> {
  const { url, token } = getConfig();
  const resp = await apiFetch(`${url}/api/space`, { headers: buildHeaders(token) });
  if (!resp.ok) throw new Error(`Failed to fetch spaces: ${resp.status}`);
  return resp.json() as Promise<TeableSpace[]>;
}

export async function listTeableBases(spaceId: string): Promise<TeableBase[]> {
  const { url, token } = getConfig();
  const resp = await apiFetch(`${url}/api/space/${encodeURIComponent(spaceId)}/base`, {
    headers: buildHeaders(token),
  });
  if (!resp.ok) throw new Error(`Failed to fetch bases: ${resp.status}`);
  return resp.json() as Promise<TeableBase[]>;
}

export async function listTeableTables(baseId: string): Promise<TeableTable[]> {
  const { url, token } = getConfig();
  const resp = await apiFetch(`${url}/api/base/${encodeURIComponent(baseId)}/table`, {
    headers: buildHeaders(token),
  });
  if (!resp.ok) throw new Error(`Failed to fetch tables: ${resp.status}`);
  return resp.json() as Promise<TeableTable[]>;
}

// ─── Schema verification & provisioning ──────────────────────────────────────

async function listFields(url: string, token: string, tableId: string): Promise<Array<{ name: string }>> {
  const resp = await apiFetch(`${url}/api/table/${encodeURIComponent(tableId)}/field`, {
    headers: buildHeaders(token),
  });
  if (!resp.ok) throw new Error(`Failed to fetch fields: ${resp.status}`);
  return resp.json() as Promise<Array<{ name: string }>>;
}

export async function verifyTeableTable(tableId: string): Promise<TeableSchemaCheck> {
  const { url, token } = getConfig();
  const fields = await listFields(url, token, tableId);
  const existing = new Set(fields.map((f) => f.name));
  const missingFields = REQUIRED_FIELDS.filter((f) => !existing.has(f.name)).map((f) => f.name);
  return { missingFields };
}

export async function ensureTeableFields(tableId: string): Promise<string[]> {
  const { url, token } = getConfig();
  const fields = await listFields(url, token, tableId);
  const existing = new Set(fields.map((f) => f.name));

  const created: string[] = [];
  for (const field of REQUIRED_FIELDS) {
    if (existing.has(field.name)) continue;

    const resp = await apiFetch(`${url}/api/table/${encodeURIComponent(tableId)}/field`, {
      method: "POST",
      headers: buildHeaders(token),
      body: JSON.stringify({ type: field.type, name: field.name }),
    });

    if (!resp.ok) {
      const body = await resp.text().catch(() => "");
      throw new Error(`Failed to create field '${field.name}': ${resp.status} — ${body}`);
    }
    created.push(field.name);
  }

  return created;
}

export async function createTeableTable(baseId: string, tableName: string): Promise<TeableTable> {
  const { url, token } = getConfig();

  const resp = await apiFetch(`${url}/api/base/${encodeURIComponent(baseId)}/table`, {
    method: "POST",
    headers: buildHeaders(token),
    body: JSON.stringify({
      name: tableName,
      fields: REQUIRED_FIELDS.map((f) => ({ type: f.type, name: f.name })),
    }),
  });

  if (!resp.ok) {
    const body = await resp.text().catch(() => "");
    throw new Error(`Failed to create table: ${resp.status} — ${body}`);
  }

  const data = await resp.json() as { id: string; name: string };
  return { id: data.id, name: data.name };
}

// ─── Lap sync ─────────────────────────────────────────────────────────────────

export async function syncLapToTeable(lap: Lap, session: LapSession): Promise<void> {
  const { url, token } = getConfig();
  const settings = getSettings();
  const tableId = settings.teableTableId;
  if (!tableId) throw new Error("No Teable table selected");

  const fields: Record<string, string | number> = {
    "Session Started At": session.startedAt,
    "Lap Mode": session.lapMode,
    "Lap Number": lap.lapNumber,
    "Start Time": new Date(lap.startTimestamp).toISOString(),
    "End Time": new Date(lap.endTimestamp).toISOString(),
    "Duration": formatDuration(lap.durationMs),
    "Speed At Start": lap.speedAtStart,
    "Speed At End": lap.speedAtEnd,
  };

  const resp = await apiFetch(`${url}/api/table/${encodeURIComponent(tableId)}/record`, {
    method: "POST",
    headers: buildHeaders(token),
    body: JSON.stringify({
      fieldKeyType: "name",
      typecast: true,
      records: [{ fields }],
    }),
  });

  if (!resp.ok) {
    const body = await resp.text().catch(() => "");
    throw new Error(`Failed to create Teable record: ${resp.status} — ${body}`);
  }
}

File: headless-rust/src/integrations/teable.rs
use crate::models::{
    SyncLapInput, TeableBase, TeableSchemaCheck, TeableSpace, TeableTable, TeableUser,
};
use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, CONTENT_TYPE};
use serde_json::json;

pub struct TeableClient {
    client: reqwest::Client,
}

impl TeableClient {
    pub fn new() -> Self {
        Self {
            client: reqwest::Client::builder()
                .danger_accept_invalid_certs(true)
                .build()
                .unwrap_or_default(),
        }
    }

    fn headers(token: &str) -> HeaderMap {
        let mut headers = HeaderMap::new();
        if let Ok(val) = HeaderValue::from_str(&format!("Bearer {}", token)) {
            headers.insert(AUTHORIZATION, val);
        }
        headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
        headers
    }

    pub async fn test_connection(&self, url: &str, token: &str) -> Result<TeableUser, String> {
        let base = url.trim_end_matches('/');
        let endpoint = format!("{}/api/user/me", base);

        let res = self
            .client
            .get(&endpoint)
            .headers(Self::headers(token))
            .send()
            .await
            .map_err(|e| e.to_string())?;

        if !res.status().is_success() {
            return Err(format!("HTTP {}", res.status()));
        }

        let user: serde_json::Value = res.json().await.map_err(|e| e.to_string())?;
        Ok(TeableUser {
            id: user["id"].as_str().unwrap_or("").to_string(),
            name: user["name"].as_str().unwrap_or("").to_string(),
            email: user["email"].as_str().unwrap_or("").to_string(),
            avatar: user["avatar"].as_str().map(|s| s.to_string()),
        })
    }

    pub async fn list_spaces(&self, url: &str, token: &str) -> Result<Vec<TeableSpace>, String> {
        let base = url.trim_end_matches('/');
        let endpoint = format!("{}/api/space", base);

        let res = self
            .client
            .get(&endpoint)
            .headers(Self::headers(token))
            .send()
            .await
            .map_err(|e| e.to_string())?;

        let json: serde_json::Value = res.json().await.map_err(|e| e.to_string())?;
        let list = json
            .as_array()
            .map(|arr| {
                arr.iter()
                    .filter_map(|s| {
                        Some(TeableSpace {
                            id: s["id"].as_str()?.to_string(),
                            name: s["name"].as_str()?.to_string(),
                        })
                    })
                    .collect()
            })
            .unwrap_or_default();

        Ok(list)
    }

    pub async fn list_bases(&self, url: &str, token: &str, space_id: &str) -> Result<Vec<TeableBase>, String> {
        let base = url.trim_end_matches('/');
        let endpoint = format!("{}/api/space/{}/base", base, space_id);

        let res = self
            .client
            .get(&endpoint)
            .headers(Self::headers(token))
            .send()
            .await
            .map_err(|e| e.to_string())?;

        let json: serde_json::Value = res.json().await.map_err(|e| e.to_string())?;
        let list = json
            .as_array()
            .map(|arr| {
                arr.iter()
                    .filter_map(|b| {
                        Some(TeableBase {
                            id: b["id"].as_str()?.to_string(),
                            name: b["name"].as_str()?.to_string(),
                            space_id: space_id.to_string(),
                        })
                    })
                    .collect()
            })
            .unwrap_or_default();

        Ok(list)
    }

    pub async fn list_tables(&self, url: &str, token: &str, base_id: &str) -> Result<Vec<TeableTable>, String> {
        let base = url.trim_end_matches('/');
        let endpoint = format!("{}/api/base/{}/table", base, base_id);

        let res = self
            .client
            .get(&endpoint)
            .headers(Self::headers(token))
            .send()
            .await
            .map_err(|e| e.to_string())?;

        let json: serde_json::Value = res.json().await.map_err(|e| e.to_string())?;
        let list = json
            .as_array()
            .map(|arr| {
                arr.iter()
                    .filter_map(|t| {
                        Some(TeableTable {
                            id: t["id"].as_str()?.to_string(),
                            name: t["name"].as_str()?.to_string(),
                        })
                    })
                    .collect()
            })
            .unwrap_or_default();

        Ok(list)
    }

    async fn list_fields(&self, url: &str, token: &str, table_id: &str) -> Result<Vec<String>, String> {
        let base = url.trim_end_matches('/');
        let endpoint = format!("{}/api/table/{}/field", base, table_id);

        let res = self
            .client
            .get(&endpoint)
            .headers(Self::headers(token))
            .send()
            .await
            .map_err(|e| e.to_string())?;

        let json: serde_json::Value = res.json().await.map_err(|e| e.to_string())?;
        let names = json
            .as_array()
            .map(|arr| {
                arr.iter()
                    .filter_map(|f| f["name"].as_str().map(|s| s.to_string()))
                    .collect()
            })
            .unwrap_or_default();

        Ok(names)
    }

    pub async fn verify_table(&self, url: &str, token: &str, table_id: &str) -> Result<TeableSchemaCheck, String> {
        let required_fields = [
            "Session Started At",
            "Lap Mode",
            "Lap Number",
            "Start Time",
            "End Time",
            "Duration",
            "Speed At Start",
            "Speed At End",
        ];

        let existing = self.list_fields(url, token, table_id).await?;
        let existing_set: std::collections::HashSet<_> = existing.into_iter().collect();

        let missing_fields = required_fields
            .iter()
            .filter(|f| !existing_set.contains(**f))
            .map(|f| f.to_string())
            .collect();

        Ok(TeableSchemaCheck { missing_fields })
    }

    pub async fn ensure_fields(&self, url: &str, token: &str, table_id: &str) -> Result<Vec<String>, String> {
        let required_fields: &[(&str, &str)] = &[
            ("Session Started At", "singleLineText"),
            ("Lap Mode", "singleLineText"),
            ("Lap Number", "number"),
            ("Start Time", "singleLineText"),
            ("End Time", "singleLineText"),
            ("Duration", "singleLineText"),
            ("Speed At Start", "number"),
            ("Speed At End", "number"),
        ];

        let existing = self.list_fields(url, token, table_id).await?;
        let existing_set: std::collections::HashSet<_> = existing.into_iter().collect();

        let base = url.trim_end_matches('/');
        let endpoint = format!("{}/api/table/{}/field", base, table_id);

        let mut created = Vec::new();
        for (name, field_type) in required_fields {
            if existing_set.contains(*name) {
                continue;
            }

            let body = json!({
                "name": name,
                "type": field_type
            });

            let res = self
                .client
                .post(&endpoint)
                .headers(Self::headers(token))
                .json(&body)
                .send()
                .await
                .map_err(|e| e.to_string())?;

            if res.status().is_success() {
                created.push(name.to_string());
            }
        }

        Ok(created)
    }

    pub async fn create_table(
        &self,
        url: &str,
        token: &str,
        base_id: &str,
        table_name: &str,
    ) -> Result<TeableTable, String> {
        let base = url.trim_end_matches('/');
        let endpoint = format!("{}/api/base/{}/table", base, base_id);

        let fields = json!([
            { "name": "Session Started At", "type": "singleLineText" },
            { "name": "Lap Mode", "type": "singleLineText" },
            { "name": "Lap Number", "type": "number" },
            { "name": "Start Time", "type": "singleLineText" },
            { "name": "End Time", "type": "singleLineText" },
            { "name": "Duration", "type": "singleLineText" },
            { "name": "Speed At Start", "type": "number" },
            { "name": "Speed At End", "type": "number" }
        ]);

        let body = json!({
            "name": table_name,
            "fields": fields
        });

        let res = self
            .client
            .post(&endpoint)
            .headers(Self::headers(token))
            .json(&body)
            .send()
            .await
            .map_err(|e| e.to_string())?;

        if !res.status().is_success() {
            let body_text = res.text().await.unwrap_or_default();
            return Err(format!("Failed to create table: {}", body_text));
        }

        let data: serde_json::Value = res.json().await.map_err(|e| e.to_string())?;
        Ok(TeableTable {
            id: data["id"].as_str().unwrap_or("").to_string(),
            name: data["name"].as_str().unwrap_or(table_name).to_string(),
        })
    }

    pub async fn sync_lap(&self, url: &str, token: &str, table_id: &str, input: &SyncLapInput) -> Result<(), String> {
        let base = url.trim_end_matches('/');
        let endpoint = format!("{}/api/table/{}/record", base, table_id);

        let start_dt = chrono::DateTime::from_timestamp_millis(input.lap.start_timestamp)
            .map(|dt| dt.to_rfc3339())
            .unwrap_or_default();

        let end_dt = chrono::DateTime::from_timestamp_millis(input.lap.end_timestamp)
            .map(|dt| dt.to_rfc3339())
            .unwrap_or_default();

        let total_sec = input.lap.duration_ms as f64 / 1000.0;
        let min = (total_sec / 60.0).floor() as i64;
        let sec = (total_sec % 60.0).floor() as i64;
        let ms = input.lap.duration_ms % 1000;
        let dur_str = format!("{:02}:{:02}.{:03}", min, sec, ms);

        let body = json!({
            "records": [{
                "fields": {
                    "Session Started At": input.session.started_at,
                    "Lap Mode": input.session.lap_mode,
                    "Lap Number": input.lap.lap_number,
                    "Start Time": start_dt,
                    "End Time": end_dt,
                    "Duration": dur_str,
                    "Speed At Start": input.lap.speed_at_start,
                    "Speed At End": input.lap.speed_at_end
                }
            }]
        });

        let res = self
            .client
            .post(&endpoint)
            .headers(Self::headers(token))
            .json(&body)
            .send()
            .await
            .map_err(|e| e.to_string())?;

        if !res.status().is_success() {
            return Err(format!("Teable sync error: HTTP {}", res.status()));
        }

        Ok(())
    }
}
